optimizer/Transformer/ServerSideRendering.php

```diff
@@ -0,0 +1,546 @@
+        DOMElement $element,
+        $layout,
+        AMP_CSS_Length $width,
+        AMP_CSS_Length $height
+    ) {
+        if ($layout !== Layout::RESPONSIVE
+            || ! $width->is_set()
+            || $width->get_numeral() === 0
+            || ! $height->is_set()
+            || $width->get_unit() !== $height->get_unit()) {
+            return;
+        }
+
+        $padding = $height->get_numeral() / $width->get_numeral() * 100;
+        $sizer   = $document->createElement(self::SIZER_ELEMENT);
+        $sizer->setAttribute('style', sprintf('display:block;padding-top:%1.4F%%;', $padding));
+        $element->insertBefore($sizer, $element->firstChild);
+    }
+
+    /**
+     * Check whether the element has an ancestor of a given tag type.
+     *
+     * @param DOMElement $element Element to check the ancestor tree of.
+     * @param string     $tagName Name of the tag to look for.
+     * @return bool Whether the element has an ancestor of the given tag name.
+     */
+    private function hasAncestorWithTag(DOMElement $element, $tagName)
+    {
+        $parent = $element->parentNode;
+        while ($parent !== null) {
+            if ($parent instanceof DOMElement && $parent->tagName === $tagName) {
+                return true;
+            }
+            $parent = $parent->parentNode;
+        }
+        return false;
+    }
+
+    /**
+     * Check if the amp-experiment element is actually used.
+     *
+     * This checks if amp-experiment has one child that is a script/json tag with a textnode that is parsable JSON and
+     * not empty. The validator ensures that the script/json is parsable but since transformers may be used outside of
+     * validation it is checked here as well.
+     *
+     * @param DOMElement $element Element to check,
+     * @return bool Whether the amp-experiment element is actually used.
+     */
+    private function isAmpExperimentUsed(DOMElement $element)
+    {
+        $script = null;
+        $child  = $element->firstChild;
+
+        while ($child) {
+            if ($child->tagName === 'script' && strtolower($child->getAttribute('type')) === 'application/json') {
+                $script = $child;
+                break;
+            }
+            $child = $child->nextSibling;
+        }
+
+        // If not script/json tag, then not used.
+        if ($script === null) {
+            return false;
+        }
+
+        // If not exactly one child is present, then not used.
+        if ($script->childNodes->length !== 1) {
+            return false;
+        }
+
+        $child = $script->firstChild;
+
+        // If child is not a text node or CDATA section, then not used.
+        if ($child->nodeType !== XML_TEXT_NODE && $child->nodeType !== XML_CDATA_SECTION_NODE) {
+            return false;
+        }
+
+        $json = $child->textContent;
+
+        // If textnode is not JSON parsable, then not used.
+        $experiment = json_decode($json, /*$assoc*/ true);
```

Comment:
What if the value is `null`? Should this not actually use `json_last_error()`?

Reply:
If the value is `null`, we'll get `null` as a result, which is still a valid reason to skip.